### PR TITLE
Jormun: Async experimental scenario

### PR DESCRIPTION
**Pull request by azime**
Support async call kraken in experimental scenario

### CODE REVIEW

File: source/jormungandr/jormungandr/scenarios/experimental.py [hunk 1]
```diff
@@ -131,16 +134,240 @@ def _reverse_journeys(res):
...
+        fallback[mode] = {}
+        return
+    import numpy as np
+    durations = np.array(sn_routing_matrix.rows[0].duration)
+    valid_duration_idx = np.argwhere((durations > -1) & (durations < max_duration)).flatten()
+    fallback[mode] = dict(zip([places[i].uri for i in valid_duration_idx],
+                    durations[(durations > -1) & (durations < max_duration)].flatten()))
+
+
+import gevent, gevent.pool
```
Comment (kinnou02) [inline source/jormungandr/jormungandr/scenarios/experimental.py]: I like import to be on the top of the file.

File: source/jormungandr/jormungandr/scenarios/experimental.py [hunk 2]
```diff
@@ -131,16 +134,240 @@ def _reverse_journeys(res):
...
+    dp_key = (dep_mode, pt_object_origin.uri, pt_object_destination.uri, request['datetime'],
+              request['clockwise'], reverse_sections)
+    dp = fallback_direct_path.get(dp_key)
+    if dp.journeys:
+        journey_parameters.direct_path_duration = dp.journeys[0].durations.total
+    else:
+        journey_parameters.direct_path_duration = None
+    origins = fall_back_origins.get(dep_mode)
+    destinations = fall_back_destinations.get(arr_mode)
+    if not fall_back_origins or not fall_back_destinations or not request.get('max_duration', 0):
```
Comment (kinnou02) [inline source/jormungandr/jormungandr/scenarios/experimental.py]: Almost everywhere we use fallback in one word not two.

File: source/jormungandr/jormungandr/scenarios/experimental.py [hunk 3]
```diff
@@ -131,16 +134,240 @@ def _reverse_journeys(res):
...
+        return futures
+
+
 class Scenario(new_default.Scenario):
 
     def __init__(self):
         super(Scenario, self).__init__()
 
-    def _get_direct_path(self, instance, mode, pt_object_origin, pt_object_destination, datetime, clockwise, request, reverse_sections=False):
+    def _get_direct_path(self, instance, mode, pt_object_origin, pt_object_destination, datetime, clockwise,
```
Comment (kinnou02) [inline source/jormungandr/jormungandr/scenarios/experimental.py]: I am a little confuse about all of this, so this function act as a cache to not compute the same direct multiple time right?

File: source/jormungandr/jormungandr/scenarios/experimental.py [hunk 4]
```diff
@@ -249,85 +408,67 @@ def call_kraken(self, request_type, request, instance, krakens_call):
...
-                                                                         reverse=True)
-
-               #Fetch all the stop points of this stop_area and replaces all the durations by 0 in the table
-               #g.destinations_fallback[arr_mode]
-               _update_crowfly_duration(instance, g.destinations_fallback, arr_mode, request['destination'],
-                                        crow_fly_stop_points)
+        worker = Worker(instance, krakens_call, request)
+
+        if request.get('max_duration', 0):
+            futures = worker._get_crowfly_futures(g.requested_origin,
```
Comment (kinnou02) [inline source/jormungandr/jormungandr/scenarios/experimental.py]: This seems to be mostly based on side effect, I don't like side effect and async, it's complex and error prone

File: source/jormungandr/jormungandr/scenarios/experimental.py [hunk 5]
```diff
@@ -249,85 +408,67 @@ def call_kraken(self, request_type, request, instance, krakens_call):
...
+                                            g.requested_destination, g.origins_places_crowfly,
+                                            g.destinations_places_crowfly, 5000)
+            gevent.joinall(futures)
+
+            futures = worker._get_routing_matrix_fustures(g.requested_origin, g.requested_destination,
```
Comment (kinnou02) [inline source/jormungandr/jormungandr/scenarios/experimental.py]: There is a lot of abstraction level and absolute not explanation about what happen, why at that moment.
I have count at least seven steps, why do we do one before the other?

**Review by kinnou02 — CHANGES_REQUESTED** [review]

File: source/jormungandr/jormungandr/scenarios/experimental.py [hunk 6]
```diff
@@ -249,90 +396,88 @@ def call_kraken(self, request_type, request, instance, krakens_call):
...
+                mode, dep_arr, resp = future.get()
+                if dep_arr == "dep":
+                    g.origins_fallback.update(resp)
+                else:
+                    g.destinations_fallback.update(resp)
+
+            futures = worker.get_duration_futures()
+            for future in gevent.iwait(futures):
+                mode, dep_arr, resp = future.get()
+                crow_fly_stop_points = set(list(crow_fly_stop_points) + resp[0])
```
Comment (xlqian) [inline source/jormungandr/jormungandr/scenarios/experimental.py]: `crow_fly_stop_points.add(resp[0])` ?

**Comment by xlqian** [comment]
Jenkins, test this please!

File: source/jormungandr/jormungandr/scenarios/experimental.py [hunk 7]
```diff
@@ -29,16 +29,20 @@
...
-from jormungandr.utils import get_uri_pt_object
+from jormungandr.utils import get_uri_pt_object, generate_id
 from jormungandr import app
-
+import gevent
+import gevent.pool
+import collections
+# http://www.gevent.org/intro.html#monkey-patching
+import gevent.monkey
+gevent.monkey.patch_socket()
```
Comment (kinnou02) [inline source/jormungandr/jormungandr/scenarios/experimental.py]: We should do the monkey patching in the init of jormun, else importing the scenario would have induce a side effect

File: source/jormungandr/jormungandr/scenarios/experimental.py [hunk 8]
```diff
@@ -244,90 +400,89 @@ def call_kraken(self, request_type, request, instance, krakens_call):
...
-               #Fetch all the stop points of this stop_area and replaces all the durations by 0 in the table
-               #g.destinations_fallback[arr_mode]
-               _update_crowfly_duration(instance, g.destinations_fallback, arr_mode, request['destination'],
-                                        crow_fly_stop_points)
+        worker = AsyncWorker(instance, krakens_call, request)
+
+        if request.get('max_duration', 0):
+            orig_futures, dest_futures = worker.get_crowfly_futures(g.requested_origin, g.requested_destination)
+            gevent.joinall(orig_futures + dest_futures)
+            for future in gevent.iwait(orig_futures):
```
Comment (kinnou02) [inline source/jormungandr/jormungandr/scenarios/experimental.py]: You don't have to do a `joinall` if you use `iwait`

File: source/jormungandr/jormungandr/scenarios/experimental.py [hunk 9]
```diff
@@ -131,16 +139,232 @@ def _reverse_journeys(res):
...
+            called_modes |= {(arr_mode, destination.uri), (dep_mode, origin.uri)}
+        return origin_futures, destination_futures
+
+    def get_crowfly_futures(self, origin, destination):
+        origin_futures = []
+        destination_futures = []
+
+        called_modes = set()
+        for dep_mode, arr_mode in self.krakens_call:
+            if (dep_mode, origin.uri) not in called_modes:
```
Comment (kinnou02) [inline source/jormungandr/jormungandr/scenarios/experimental.py]: We don't need the `origin.uri` here no?

File: source/jormungandr/jormungandr/scenarios/experimental.py [hunk 10]
```diff
@@ -244,90 +400,89 @@ def call_kraken(self, request_type, request, instance, krakens_call):
...
-                                                                         reverse=True)
-
-               #Fetch all the stop points of this stop_area and replaces all the durations by 0 in the table
-               #g.destinations_fallback[arr_mode]
-               _update_crowfly_duration(instance, g.destinations_fallback, arr_mode, request['destination'],
-                                        crow_fly_stop_points)
+        worker = AsyncWorker(instance, krakens_call, request)
+
+        if request.get('max_duration', 0):
+            orig_futures, dest_futures = worker.get_crowfly_futures(g.requested_origin, g.requested_destination)
```
Comment (kinnou02) [inline source/jormungandr/jormungandr/scenarios/experimental.py:501]: comment to add:
First we search all stop_point reachable for each modes from the origin and the destination

File: source/jormungandr/jormungandr/scenarios/experimental.py [hunk 11]
```diff
@@ -244,90 +400,89 @@ def call_kraken(self, request_type, request, instance, krakens_call):
...
+            gevent.joinall(orig_futures + dest_futures)
+            for future in gevent.iwait(orig_futures):
+                g.origins_places_crowfly.update(future.get())
+            for future in gevent.iwait(dest_futures):
+                g.destinations_places_crowfly.update(future.get())
+
+            orig_futures, dest_futures = worker.get_routing_matrix_futures(g.requested_origin,
```
Comment (kinnou02) [inline source/jormungandr/jormungandr/scenarios/experimental.py:516]: Comment to add:
Then we want to put a arrival time for each of this stop points.
As a side note this won't work the day when our ETA will be impacted by the datetime of the journey, at least for the arrival when doing a "departure after" request.

File: source/jormungandr/jormungandr/scenarios/experimental.py [hunk 12]
```diff
@@ -244,90 +400,89 @@ def call_kraken(self, request_type, request, instance, krakens_call):
...
+                                                                           g.requested_destination,
+                                                                           g.origins_places_crowfly,
+                                                                           g.destinations_places_crowfly)
+            gevent.joinall(orig_futures + dest_futures)
+            for future in gevent.iwait(orig_futures):
+                g.origins_fallback.update(future.get())
+            for future in gevent.iwait(dest_futures):
+                g.destinations_fallback.update(future.get())
+
+            orig_futures, dest_futures = worker.get_update_crowfly_duration_futures()
```
Comment (kinnou02) [inline source/jormungandr/jormungandr/scenarios/experimental.py:528]: If I understand correctly we update the stop_points of the requested stop_area to set the ETA at 0, if so this call to kraken can have be made earlier.

File: source/jormungandr/jormungandr/scenarios/experimental.py [hunk 13]
```diff
@@ -244,90 +400,89 @@ def call_kraken(self, request_type, request, instance, krakens_call):
...
-                if direct_path not in resp:
-                    direct_path.journeys[0].internal_id = "dp_{resp}-{j}".format(resp=self.nb_kraken_calls, j=0)
-                    resp.append(direct_path)
-            else:
-                journey_parameters.direct_path_duration = None
 
-            fall_back_origins = g.origins_fallback.get(dep_mode)
-            fall_back_destinations = g.destinations_fallback.get(arr_mode)
-            if not fall_back_origins or not fall_back_destinations or not request.get('max_duration', 0):
+        futures = worker.get_direct_path_futures(self._get_direct_path,
```
Comment (kinnou02) [inline source/jormungandr/jormungandr/scenarios/experimental.py]: Comment to add:
we compute the direct path for this journey, their time will be used to initialized our PT calls

File: source/jormungandr/jormungandr/scenarios/experimental.py [hunk 14]
```diff
@@ -244,90 +400,89 @@ def call_kraken(self, request_type, request, instance, krakens_call):
...
+                                                 g.requested_destination)
+        for future in gevent.iwait(futures):
+            self.nb_kraken_calls += 1
+            dep_mode, resp_direct_path = future.get()
+            if resp_direct_path.journeys:
+                if resp_direct_path not in resp:
+                    resp_direct_path.journeys[0].internal_id = str(generate_id())
+                    resp.append(resp_direct_path)
+
+        futures = worker.get_journey_futures(g.requested_origin, g.requested_destination,
```
Comment (kinnou02) [inline source/jormungandr/jormungandr/scenarios/experimental.py]: Comment to add:
We launch the PT journey planning

This calls could be launched earlier, as soon as we have the direct_path duration for the requested mode

File: source/jormungandr/jormungandr/scenarios/experimental.py [hunk 15]
```diff
@@ -244,90 +400,89 @@ def call_kraken(self, request_type, request, instance, krakens_call):
...
                 continue
-
-            local_resp = instance.planner.journeys(fall_back_origins,
-                                                   fall_back_destinations,
-                                                   request['datetime'],
-                                                   request['clockwise'],
-                                                   journey_parameters)
+            dp_key = (dep_mode, g.requested_origin.uri, g.requested_destination.uri, request['datetime'],
+                      request['clockwise'], False)
+            direct_path = g.fallback_direct_path.get(dp_key)
```
Comment (kinnou02) [inline source/jormungandr/jormungandr/scenarios/experimental.py:573]: I didn't see where we compute the fallback direct_path

File: source/jormungandr/jormungandr/scenarios/experimental.py [hunk 16]
```diff
@@ -244,90 +400,89 @@ def call_kraken(self, request_type, request, instance, krakens_call):
...
-                                    request)
-
+            for j in local_resp.journeys:
+                j.internal_id = str(generate_id())
+                map_response.append((dep_mode, arr_mode, j))
             resp.append(local_resp)
-            logger.debug("for mode %s|%s we have found %s journeys", dep_mode, arr_mode, len(local_resp.journeys))
-            logger.debug("for mode %s|%s we have found %s direct path", dep_mode, arr_mode, len(direct_path.journeys))
+
+        futures = worker.build_journeys(self._extend_journey, map_response, crow_fly_stop_points)
```
Comment (kinnou02) [inline source/jormungandr/jormungandr/scenarios/experimental.py]: Comment to add:
We compute the fallback part of our PT journeys

File: source/jormungandr/jormungandr/scenarios/experimental.py [hunk 17]
```diff
@@ -244,90 +400,89 @@ def call_kraken(self, request_type, request, instance, krakens_call):
...
+        for future in gevent.iwait(futures):
```
Comment (kinnou02) [inline source/jormungandr/jormungandr/scenarios/experimental.py:592]: Why not do that direclty in `build_journey`?

**Review by kinnou02 — CHANGES_REQUESTED** [review]

File: source/jormungandr/jormungandr/scenarios/experimental.py (hunk 7 above)
Comment (xlqian) [reply source/jormungandr/jormungandr/scenarios/experimental.py]: done

File: source/jormungandr/jormungandr/scenarios/experimental.py (hunk 9 above)
Comment (xlqian) [reply source/jormungandr/jormungandr/scenarios/experimental.py]: I have simplified that logic :)

File: source/jormungandr/jormungandr/scenarios/experimental.py (hunk 11 above)
Comment (xlqian) [reply source/jormungandr/jormungandr/scenarios/experimental.py:516]: Yep, I'm adding a lot of comments :)

File: source/jormungandr/jormungandr/scenarios/experimental.py (hunk 8 above)
Comment (xlqian) [reply source/jormungandr/jormungandr/scenarios/experimental.py]: I imagined a scenario as followed: 

We need to compute some crowfly stop_points with "walking" and "bike" respectively for "from" and "to", in this case we'll get 4 requests for stop_points.

**WITH** `joinall`: all requests are launched at the same time thanks to the concatenation
**WITHOUT** `joinall`: only 2 requests are parallelized 

That's why i did with `joinall`

File: source/jormungandr/jormungandr/scenarios/experimental.py (hunk 12 above)
Comment (xlqian) [reply source/jormungandr/jormungandr/scenarios/experimental.py:528]: Hmm, this functions update the sn_routing_matrix if the stop_points is contained by stop_area, so we can only do this after the computation of sn_routing_matrix

File: source/jormungandr/jormungandr/scenarios/experimental.py (hunk 14 above)
Comment (xlqian) [reply source/jormungandr/jormungandr/scenarios/experimental.py]: `This calls could be launched earlier, as so......`
Nope, we need the `global` direct_path duration to launch this pt computation

File: source/jormungandr/jormungandr/scenarios/experimental.py (hunk 15 above)
Comment (xlqian) [reply source/jormungandr/jormungandr/scenarios/experimental.py:573]: my bad, i'm improving this

File: source/jormungandr/jormungandr/scenarios/experimental.py (hunk 3 above)
Comment (xlqian) [reply source/jormungandr/jormungandr/scenarios/experimental.py]: I'm improving this (which is not that easy)

File: source/jormungandr/jormungandr/scenarios/experimental.py (hunk 8 above)
Comment (xlqian) [reply source/jormungandr/jormungandr/scenarios/experimental.py]: I removed the iwait :)

File: source/jormungandr/jormungandr/scenarios/experimental.py [hunk 18]
```diff
@@ -265,71 +477,98 @@ def call_kraken(self, request_type, request, instance, krakens_call):
...
+        # Now we compute the direct path with all requested departure mode
+        # their time will be used to initialized our PT calls
+        futures = worker.get_direct_path_futures(g.fallback_direct_path,
+                                                 g.requested_origin,
+                                                 g.requested_destination,
+                                                 request['datetime'],
+                                                 request['clockwise'],
+                                                 False)
+        for future in gevent.iwait(futures):
+            self.nb_kraken_calls += 1
```
Comment (azime) [inline source/jormungandr/jormungandr/scenarios/experimental.py]: to delete ?

**Comment by xlqian** [comment]
I benched on `new_default` w/wo `gevet.monkey.patch_socket()` There is no apparent regression in performance.

**Review by kinnou02 — APPROVED** [review]

**Pull request merged**